### PR TITLE
Fix recursive package require

### PR DESCRIPTION
**Pull request by renan028**
1) fix infinite loop when packages require each other (or conflict with each other):
A requires B
B requires C

2) workaround when version specification is not a number (just ignores it for now)

### CODE REVIEW

File: lib/utilrb/pkgconfig.rb [hunk 1]
```diff
@@ -207,16 +216,33 @@ def self.parse_dependencies(string)
...
+            
+            if memo.empty?
+                depth = 1
+            else
+                depth = memo.values.max + 1
+            end
+
+            packages.map do |dep|
+                dep = dep.strip
+                if !memo.key?(dep)
```
Comment (doudou) [inline lib/utilrb/pkgconfig.rb]: `if !` is written `unless`, and move to a one-line form for small one-line conditionals, e.g.

```
memo[dep] = depth unless memo[dep]
```

However, in this case, since values are not `nil`, the idiomatic thing to write would be

```
memo[dep] ||= depth
```

File: lib/utilrb/pkgconfig.rb [hunk 2]
```diff
@@ -192,7 +201,7 @@ def perform_substitution(value, variables, current)
             value
         end
 
-        def self.parse_dependencies(string)
+        def self.parse_dependencies(string, memo:Hash.new)
```
Comment (doudou) [inline lib/utilrb/pkgconfig.rb]: Style: space betweem `:` and `Hash`

File: lib/utilrb/pkgconfig.rb [hunk 3]
```diff
@@ -74,13 +74,17 @@ def self.get(name, version_spec = nil, preset_variables = Hash.new, minimal: fal
 
             # Now try to find a matching spec
             if match = find_matching_version(candidates, version_spec)
-                match
+                if version_spec.eql? match
+                    return nil
```
Comment (doudou) [inline lib/utilrb/pkgconfig.rb]: One usually does not return `nil` explicitely. Just `return` is enough.

File: lib/utilrb/pkgconfig.rb [hunk 4]
```diff
@@ -74,13 +74,17 @@ def self.get(name, version_spec = nil, preset_variables = Hash.new, minimal: fal
...
+                else
+                    match
```
Comment (doudou) [inline lib/utilrb/pkgconfig.rb]: This `match` - which was there before - looks like a leftover from old copy/paste code. It's useless.

File: lib/utilrb/pkgconfig.rb [hunk 5]
```diff
@@ -74,13 +74,17 @@ def self.get(name, version_spec = nil, preset_variables = Hash.new, minimal: fal
 
             # Now try to find a matching spec
             if match = find_matching_version(candidates, version_spec)
-                match
+                if version_spec.eql? match
```
Comment (doudou) [inline lib/utilrb/pkgconfig.rb]: This looks very weird ... What is it doing ?

File: lib/utilrb/pkgconfig.rb [hunk 6]
```diff
@@ -207,16 +216,33 @@ def self.parse_dependencies(string)
...
+            packages.map do |dep|
+                dep = dep.strip
+                if !memo.key?(dep)
+                    memo[dep] = depth
+                end
+            end
 
             result = packages.map do |dep|
                 dep = dep.strip
+                if memo[dep] < depth
```
[the remark this answers is not in the document]
Comment (doudou) [reply lib/utilrb/pkgconfig.rb]: Why this test ? It looks weird ...

File: test/test_pkgconfig.rb [hunk 7]
```diff
@@ -166,4 +166,14 @@ def test_requires
         assert_equal ['test_pkgconfig'],
             pkg.requires.map(&:name)
     end
+
+    def test_recursive_requires
```
Comment (doudou) [inline test/test_pkgconfig.rb]: You don't have a test for the malformed version field.

File: lib/utilrb/pkgconfig.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (doudou) [reply lib/utilrb/pkgconfig.rb]: > Do you think we should provide the full dependencies for required packages also ?

Yes. I definitely expect each and every object in the chain to be well-formed.

As we discussed, dropping the depth thing to simply use the `memo` as a cache (use an object in `memo` and resolve it otherwise) would actually be a simpler implementation, *and* would provide "proper" PkgCOnfig objects in the complete dependency chain

File: lib/utilrb/pkgconfig.rb [hunk 8]
```diff
@@ -376,10 +347,10 @@ def load_fields(memo)
             # And finally resolve the compilation flags
             @cflags = fields['Cflags'] || []
             @requires.each do |pkg|
-                @cflags.concat(pkg.raw_cflags)
+                @cflags.concat(pkg.raw_cflags) unless pkg.raw_cflags.nil?
```
Comment (doudou) [inline lib/utilrb/pkgconfig.rb]: Why do you need the guard now, but did not need it before ?

File: test/test_pkgconfig.rb [hunk 9]
```diff
@@ -168,12 +168,27 @@ def test_requires
...
+        assert_equal ['test_pkgconfig_recursive_require_loop_b', 'test_pkgconfig_recursive_require_loop_c'],
+            pkg.requires[0].requires.map(&:name)
+    
+    end
+
+    def test_version_not_number
+        pkg = Utilrb::PkgConfig.parse_dependencies('test_pkgconfig_version_not_number')[0]
+        assert_equal ['test_pkgconfig_recursive_require_loop_c', 'test_pkgconfig_recursive_require_loop_d'],
+            pkg.requires_private.map(&:name)
     end
```
Comment (doudou) [inline test/test_pkgconfig.rb]: You had this question about whether requirements should have a "real" list of requires themselves.

If you asked yourself the question, write a test to verify the behavior

File: test/test_pkgconfig.rb [hunk 10]
```diff
@@ -168,12 +168,27 @@ def test_requires
...
+        pkg = Utilrb::PkgConfig.parse_dependencies('test_pkgconfig_recursive_require_loop_b')[0]
+        assert_equal ['test_pkgconfig_recursive_require_loop_a', 'test_pkgconfig_recursive_require_loop_b', 'test_pkgconfig_recursive_require_loop_d'],
+            pkg.requires.map(&:name)
+            
+        assert_equal ['test_pkgconfig_recursive_require_loop_b', 'test_pkgconfig_recursive_require_loop_c'],
+            pkg.requires[0].requires.map(&:name)
+    
+    end
+
+    def test_version_not_number
```
Comment (doudou) [inline test/test_pkgconfig.rb]: There's one edge-case that I think you are already covering, but I don't know for sure. Best would be to write a test for it: `A >= B >= 0.2`
